### PR TITLE
implement remaining functions in `mlx.nn.py` excluding pooling

### PR DESCRIPTION
**Pull request by acsweet**
This addresses #19571, to complete the rest of `mlx.nn.py` excluding the pooling functions

This implements:
- `psnr`
- `ctc_loss`
- `ctc_decode`
    - `_ctc_greedy_decode`
    - `_ctc_beam_search_decode`
        - it's not very efficient, mostly because mlx does not have a `unique` function (I probably spent too much time on this 😅)
        - we could disable it in the tests (I saw that for torch)

This patches the following:
- `one_hot`
- `categorical_crossentropy`
- `moments` 
- `batch_normalization`
    - this now depends on jax's `norm` function until mlx's `norm` supports singular value norms
    - (is numpy preferred here?)

### CODE REVIEW

**Comment by fchollet** [comment]
Impressive that you managed to implement `ctc_decode` -- very nice work.

> is numpy preferred here

There is not big difference between relying on jax or np here, jax is likely to be somewhat more performant but the different should not be material. Using jax is fine.

**Comment by fchollet** [comment]
> excluding the pooling functions

@fbadine are you able to add the pooling functions?

**Review by fchollet — APPROVED** [review]
LGTM

**Comment by fbadine** [comment]
> > excluding the pooling functions
> 
> 
> 
> @fbadine are you able to add the pooling functions?

Working on them. I should be able to finish them in the next couple of days.

**Comment by fbadine** [comment]
> > > excluding the pooling functions
> > 
> > 
> > 
> > @fbadine are you able to add the pooling functions?
> 
> Working on them. I should be able to finish them in the next couple of days. 

@fchollet, the code is almost done however the average pooling with same padding is failing with values similar to torch values and different from tensorflow and jax. 
For torch I saw that we skipped this test, should we do the same for mlx? The 2 are similar in terms of having to calculate the padding values manually.

This is from torch

keras/src/ops/nn_test.py::NNOpsCorrectnessTest::test_average_pool_same_padding SKIPPED (Torch outputs differently from TF when using `same` padding.)

**Comment by acsweet** [comment]
@fbadine I had to handle the `same` padding in a custom way for `conv`, take a look at `compute_transpose_padding_args_for_mlx` in `keras/src/backend/backend_utils.py`

Also I think I named that wrong, we can rename it to `compute_conv_padding_args_for_mlx` or if it works for the pooling layers, we can just call it `compute_padding_args_for_mlx`

**Comment by fbadine** [comment]
Actually it turned out to be related to average pooling counting the padding into the average which led to different values. 
I applied then same logic as in JAX implementation and now the tests are passing.
I will review and clean up the code and will open a PR for pooling
